gateway/app/drivers/quality_on_demand/nef/nef_interface.py

```diff
@@ -0,0 +1,359 @@
+    SessionInfo,
+    CloudEvent,
+    Specversion,
+    StatusInfo,
+    Type,
+)
+from app.settings import settings
+
+_prefix_info = "qodinfo"
+_prefix_gateway_nef = "qodnef"
+_prefix_device = "qodDevice"
+LOG = logging.getLogger(__name__)
+
+
+class NEFQoDInterface(QoDInterface):
+    def __init__(self, nef_url: AnyHttpUrl, nef_auth: NEFAuth) -> None:
+        super().__init__()
+        self.httpx_client = httpx.AsyncClient(base_url=str(nef_url), auth=nef_auth)
+        self.httpx_client_callback = httpx.AsyncClient()
+        self.redis = get_redis()
+
+    async def create_provisioning(
+        self, req: CreateSession, device: Device
+    ) -> SessionInfo:
+        qod_id = uuid.uuid4()
+
+        payload = AsSessionWithQoSSubscription(
+            supportedFeatures="0",
+            notificationDestination=Link(
+                f"{settings.nef_gateway_url}callbacks/v1/qod/{qod_id}"
+            ),
+            flowInfo=[
+                FlowInfo(
+                    flowId=0,
+                    flowDescriptions=self._get_flow_description(req),
+                )
+            ],
+            qosReference=req.qosProfile,
+            usageThreshold=UsageThreshold(
+                duration=req.duration, totalVolume=0, downlinkVolume=0, uplinkVolume=0
+            ),
+        )
+
+        if device.ipv4Address and device.ipv4Address.publicAddress:
+            payload.ueIpv4Addr = device.ipv4Address.publicAddress
+        elif device.ipv6Address:
+            payload.ueIpv6Addr = device.ipv6Address
+        elif device.phoneNumber:
+            phone_number = device.phoneNumber.lstrip("+")
+            payload.gpsi = f"msisdn-{phone_number}"
+
+        res = await self.httpx_client.post(
+            f"/nef/api/v1/3gpp-as-session-with-qos/v1/{settings.qod_provisioning.af_id}/subscriptions",
+            content=payload.model_dump_json(exclude_unset=True),
+            headers={"Content-Type": "application/json"},
+        )
+
+        if res.status_code == 404:
+            raise ResourceNotFound()
+
+        if res.status_code == 409:
+            raise SessionConflict()
+
+        if not res.is_success:
+            raise BadRequest()
+
+        subcription_result = AsSessionWithQoSSubscription.model_validate_json(
+            res.content
+        )
+
+        nef_sub_id = str(subcription_result.self)
+
+        response = SessionInfo(
+            sessionId=qod_id,
+            device=device,
+            applicationServer=req.applicationServer,
+            devicePorts=req.devicePorts,
+            applicationServerPorts=req.applicationServerPorts,
+            qosProfile=req.qosProfile,
+            duration=req.duration,
+            sink=req.sink,
+            startedAt=datetime.datetime.now(),
+            expiresAt=datetime.datetime.now()
+            + datetime.timedelta(seconds=req.duration),
+            qosStatus=QosStatus.REQUESTED,
+        )
+        await self._save_subscription_info(str(qod_id), response)
```

Comment:
Create the gateway subscription before the NEF subscription. This way NEF notifications will always have access to the gateway subscription.